lib/view_component/slotable_v2.rb

```diff
@@ -0,0 +1,222 @@
+        # helper method with the same name as the slot. For collection
+        # components, the method can be called multiple times to append to the
+        # slot.
+        #
+        #   <%= render_inline(MyComponent.new) do |component| %>
+        #     <%= component.header(classes: "Foo") do %>
+        #       <p>Bar</p>
+        #     <% end %>
+        #   <% end %>
+        def renders_one(slot_name, callable = nil)
+          validate_slot_name(slot_name)
+
+          define_method slot_name do |*args, **kwargs, &block|
+            if args.empty? && kwargs.empty? && block.nil?
+              get_slot(slot_name)
+            else
+              set_slot(slot_name, *args, **kwargs, &block)
+            end
+          end
+
+          register_slot(slot_name, collection: false, callable: callable)
+        end
+
+        ##
+        # Registers a collection slot on the component.
+        #
+        # = Example
+        #
+        #   render_many :items do
+        #     def initialize(name:)
+        #       @name = name
+        #     end
+        #   end
+        #
+        # = Rendering slot content
+        #
+        # The component's sidecar template can access the slot by calling a
+        # helper method with the same name as the slot.
+        #
+        #   <h1>
+        #     <%= items.each do |item| %>
+        #       <%= item %>
+        #     <% end %>
+        #   </h1>
+        #
+        # = Setting slot content
+        #
+        # Renderers of the component can set the content of a slot by calling a
+        # helper method with the same name as the slot. The method can be
+        # called multiple times to append to the slot.
+        #
+        #   <%= render_inline(MyComponent.new) do |component| %>
+        #     <%= component.item(name: "Foo") do %>
+        #       <p>One</p>
+        #     <% end %>
+        #
+        #     <%= component.item(name: "Bar") do %>
+        #       <p>two</p>
+        #     <% end %>
+        #   <% end %>
+        def renders_many(slot_name, callable = nil)
+          validate_slot_name(slot_name)
+
+          singular_name = ActiveSupport::Inflector.singularize(slot_name)
+
+          # Define setter for singular names
+          # e.g. `with_slot :tab, collection: true` allows fetching all tabs with
+          # `component.tabs` and setting a tab with `component.tab`
+          define_method singular_name do |*args, **kwargs, &block|
+            # TODO raise here if attempting to get a collection slot using a singular method name?
+            # e.g. `component.item` with `with_slot :item, collection: true`
+            set_slot(slot_name, *args, **kwargs, &block)
+          end
+
+          # Instantiates and and adds multiple slots forwarding the first
+          # argument to each slot constructor
+          define_method slot_name do |*args, **kwargs, &block|
+            if args.empty? && kwargs.empty? && block.nil?
```

Comment:
Let's investigate implementing a collection setter in this PR: `component.items(@items)`, as it may affect this method.